qiskit_nature/second_q/operators/polynomial_tensor.py

```diff
@@ -740,12 +741,20 @@ def einsum(
         Returns:
             A new ``PolynomialTensor``.
         """
-        dense_operands = [op.to_dense() for op in operands]
+        if _optionals.HAS_OPT_EINSUM:
+            # pylint: disable=import-error
+            from opt_einsum import contract
```

Comment:
Do we always want to use when installed or just if the PT is not dense (i.e. has sparse array(s) in it). Elsewhere we use numpy einsum - so this is really only for this local use, because of sparse, unless things were changed to say have a more central einsum that worked either way that all the other usages could switch over to. `dense_operands` seems a bit misleading as name now - maybe more simply say `einsum_operands` as in the opt_einsum case these can be sparse.

Should we have/update a unit test covering this new path to assert when installed that the result if things were sparse going into the einsum it's sparse coming out of it too (as it stood before they would all be dense right - the PT equality test does not care about that aspect hence we might want to add a check to ensure things are as we expect when opt_einsum is sued.

Reply:
I think if `opt_einsum` is available, we should always use it. Their docs state that `np.einsum(..., optimize=True)` is the upstreamed version of a lot of the features originally implemented by `opt_einsum`. But that their package still provides more up-to-date algorithms ([source](https://optimized-einsum.readthedocs.io/en/stable/#features)).

I agree that `dense_operands` is misleading now. Maybe we can just rename it to `_operands` or `operand_list` or similar.

I agree that the unittests should also be updated to test this.

Reply:
> I think if opt_einsum is available, we should always use it.

At present its only used here in PT right, where it was wanted to handle sparse better - we have other places numpy einsum is used, but there we do not use sparse at all. If we are using it on dense here too, if installed, did we try checking performance on dense - if its faster/more performant maybe it would be useful elsewhere too?